crescent/context/utils.py

```diff
@@ -0,0 +1,74 @@
+from __future__ import annotations
+
+from functools import wraps
+from typing import TYPE_CHECKING, Awaitable, cast
+
+from crescent.context.base_context import BaseContext
+from crescent.context.context import Context
+
+if TYPE_CHECKING:
+    from inspect import Parameter
+    from typing import Any, Callable, Sequence, TypeVar
+
+    from typing_extensions import ParamSpec
+
+    T = TypeVar("T")
+    P = ParamSpec("P")
+
+
+from crescent.utils import get_parameters
+
+__all__: Sequence[str] = ("support_custom_context", "get_function_context", "get_context_type")
+
+
+def _get_ctx(args: Sequence[Any]) -> tuple[BaseContext, int]:
+    """Get a variable of type `BaseContext` from a list."""
+    for index, arg in enumerate(args):
+        if isinstance(arg, BaseContext):
+            return arg, index
+    raise ValueError("Args do not include a `BaseContext` object.")
+
+
+def support_custom_context(
+    func: Callable[P, Awaitable[T]]
+) -> Callable[P, Awaitable[tuple[T, BaseContext]]]:
+    """
+    Returns a function that is called with the context type it is annotated with.
+    """
+
+    ctx_t = get_function_context(func)
+
+    @wraps(func)
+    async def inner(*args: P.args, **kwargs: P.kwargs) -> tuple[T, BaseContext]:
+
+        ctx, index = _get_ctx(args)
+
+        argv: Sequence[Any]
+        if type(ctx) is not ctx_t:
+            ctx = ctx.into(ctx_t)
+            argv = list(args)
+            argv[index] = ctx
+        else:
+            argv = args
+
+        return await func(*argv, **kwargs), ctx  # pyright: ignore
+
+    return inner
```

Comment:
Is there a reason you don't always assume a command has a custom context? If it's to make sure that you don't do unecessary checks, why not just check if the context annotation is a subclass and use that to determine whether or not to do extra steps?

It would be annoying to me to have to add this decorator to every command (or even as a kwarg).

Reply:
I do assume every command has a custom context

Reply:
Oh okay. Clearly I didn't read the changes well enough lol